src/Protocol/Protocol_1_9.cpp

```diff
@@ -2870,6 +2881,55 @@ void cProtocol_1_9_0::HandleVanillaPluginMessage(cByteBuffer & a_ByteBuffer, con
 		m_Client->HandleNPCTrade(SlotNum);
 		return;
 	}
+	else if ((a_Channel == "MC|BSign") || (a_Channel == "MC|BEdit"))
+	{
+		HANDLE_READ(a_ByteBuffer, ReadBEInt16, Int16, ItemID);
+		if (ItemID != E_ITEM_BOOK_AND_QUILL)
+		{
+			// Item is not a writeable book
+			return;
+		}
+
+		// Skip item count (1 byte) and item damage (2 bytes)
+		a_ByteBuffer.SkipRead(3);
+
+		// Read nbt content
+		AString BookData;
+		a_ByteBuffer.ReadString(BookData, a_ByteBuffer.GetReadableSpace() - 1);
+		cParsedNBT NBT(BookData.c_str(), BookData.size());
+
+		cItem BookItem;
+		if (a_Channel == "MC|BSign")
+		{
+			BookItem = cItem(E_ITEM_WRITTEN_BOOK);
+			// Add the text to a json string
+			cBookContent::ParseFromNBT(0, BookItem.m_BookContent, NBT, true);
+		}
+		else
+		{
+			BookItem = cItem(E_ITEM_BOOK_AND_QUILL);
+			cBookContent::ParseFromNBT(0, BookItem.m_BookContent, NBT);
+		}
+
+		cPlayer * Player = m_Client->GetPlayer();
+
+		// If true, player has clicked on the sign button
+		bool IsSigned = (BookItem.m_ItemType == E_ITEM_WRITTEN_BOOK) ? true : false;
+
+		if (cRoot::Get()->GetPluginManager()->CallHookPlayerEditingBook(*Player, BookItem.m_BookContent, IsSigned))
+		{
+			// Plugin denied the editing of the book
+			return;
+		}
+
+		// Book has been edited, inform plugins
+		cRoot::Get()->GetPluginManager()->CallHookPlayerEditedBook(*Player, BookItem.m_BookContent, IsSigned);
+
+		cInventory & inv = Player->GetInventory();
+		inv.SetHotbarSlot(inv.GetEquippedSlotNum(), BookItem);
+		SendWholeInventory(*Player->GetWindow());  // TODO: Use SendSlot
```

Comment:
Found this one... `Player.GetInventory().SendEquippedSlot();`

Should the handling of the books moved into cClientHandle / cPlayer?